### PR TITLE
PR for version 1

### PR DESCRIPTION
**Pull request by arunmcherian94**

### CODE REVIEW

File: .env.template [hunk 1]
```diff
@@ -0,0 +1,61 @@
+
+# coding: utf-8
+from decouple import config
```
Comment (Shwetabhk) [inline .env.template:3]: would prefer `setup.sh.template` instead of this, as we'll be using environment variables for deploy

File: .env.template [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+from unipath import Path
+from dj_database_url import parse as db_url
+
+
+BASE_DIR = Path(__file__).parent
```
Comment (Shwetabhk) [inline .env.template:8]: Why is this needed?

File: apps/backups/models.py [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+
+class Backups(models.Model):
+    """
+    Table to store user's backup request
+    """
+    id = models.AutoField(primary_key=True)
+    fyle_org_id = models.CharField(max_length=255, help_text='Fyle org_id of backup requester')
+    fyle_refresh_token = models.CharField(max_length=512, help_text='Fyle Refresh Token')
+    user = models.ForeignKey(UserProfile, on_delete=models.CASCADE)
```
Comment (Shwetabhk) [inline apps/backups/models.py:23]: Please use `models.PROTECT` do not `CASCADE DELETE`

File: apps/backups/models.py [hunk 4]
```diff
@@ -0,0 +1,43 @@
+from django.db import models
+from apps.user.models import UserProfile
+
+
+class ObjectLookup(models.Model):
```
Comment (Shwetabhk) [inline apps/backups/models.py:5]: I don't think we need a separate table for this, just Enums will do
please check out model Choices in Django
https://docs.djangoproject.com/en/3.0/ref/models/fields/

File: apps/backups/models.py [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+
+class Backups(models.Model):
+    """
+    Table to store user's backup request
+    """
+    id = models.AutoField(primary_key=True)
+    fyle_org_id = models.CharField(max_length=255, help_text='Fyle org_id of backup requester')
+    fyle_refresh_token = models.CharField(max_length=512, help_text='Fyle Refresh Token')
+    user = models.ForeignKey(UserProfile, on_delete=models.CASCADE)
+    object_type = models.ForeignKey(ObjectLookup, on_delete=models.CASCADE)
```
Comment (Shwetabhk) [inline apps/backups/models.py:24]: separate table not needed

File: apps/backups/models.py [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+    current_state = models.CharField(max_length=64, help_text="Current state of backup")
+    name = models.CharField(max_length=64, help_text="Backup name")
+    task_id = models.CharField(max_length=255, null=True,
+                               help_text='Task reference for Fyle Jobs Infra')
+    filters = models.TextField(help_text='Request URL with filters and constraints')
```
Comment (Shwetabhk) [inline apps/backups/models.py:29]: by filter do we mean the configuration of the backup? if yes can you share an example of how we're storing this

File: apps/backups/templates/expenses.html [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+                    <select class="selectpicker" multiple data-live-search="true"
+                        name="state" id="id_state">
+                                {% for choice, value in form.state.field.choices %}
+                                    <option value={{choice}}>{{ value }}</option>
+                                {% endfor %}
+                    </select>
+                </div>
+            </div>
+            <div class="filter-row">
+                <label class="filter-lbl">Approved On</label>
```
Comment (Shwetabhk) [inline apps/backups/templates/expenses.html:38]: Approved on should only be active for 
`APPROVED`,
`PAYMENT_PENDING`
`PAYMENT_PROCESSING`,
`PAID`
`ALL`

File: apps/backups/urls.py [hunk 8]
```diff
@@ -0,0 +1,11 @@
+from django.urls import path
+
+from . import views
+
+urlpatterns = [
+    path('home/', views.HomeView.as_view(), name='backups-home'),
+    path('callback/', views.OAuthCallbackView.as_view(), name='backups-callback'),
+    path('expenses/', views.ExpensesView.as_view(), name='backups-expenses'),
+    path('backups/', views.BackupsView.as_view(), name='backups-backup'),
+    path('backups/notify/<int:backup_id>', views.BackupsNotifyView.as_view(), name='backups-notify')
```
Comment (Shwetabhk) [inline apps/backups/urls.py:10]: please end with a trailing slash

File: apps/backups/views.py [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+                                                      'error_message', 'created_at')[:5]
+        return JsonResponse({"backups": list(backups_list)})
+
+    def post(self, request):
+        try:
+            form = ExpenseForm(request.POST)
+            logger.info('Got a backup request from %s with params: %s',
+                        request.user, request.POST)
+            if form.is_valid():
+                data = form.cleaned_data
```
Comment (Shwetabhk) [inline apps/backups/views.py:74]: all this logic should not be in the views, the views should only be responsible for
* Catering requests
* Calling a function that does the heavy lifting
* sending back the response from the function

File: apps/data_fetcher/views.py [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+        try:
+            backup = Backups.objects.get(id=backup_id)
+        except Backups.DoesNotExist:
+            logger.error('Invalid backup_id sent by JobsInfra. Request: %s', request.POST)
+            return HttpResponse(status=400)
+
+        filters = json.loads(backup.filters)
+        download_attachments = filters.get('download_attachments')
+        refresh_token = backup.fyle_refresh_token
+        fyle_org_id = backup.fyle_org_id
```
Comment (Shwetabhk) [inline apps/data_fetcher/views.py:35]: all the logic code should be out of the views

File: apps/backups/views.py [hunk 11]
```diff
@@ -0,0 +1,164 @@
...
+    Insert/Get objects from backups table
+    """
+    def get(self, request, object_type=None):
+        if object_type is None:
+            return {'backups': None}
+        backups_list = Backups.objects.filter(object_type_id__name=object_type,
+                                              user_id__email=request.user
+                                             ).values('id', 'name',
+                                                      'current_state',
+                                                      'error_message', 'created_at')[:5]
```
Comment (aditya999123) [inline apps/backups/views.py:65]: why [:5] ?
and even if you want to limit this, use env var, so that this is configurable

File: apps/backups/views.py [hunk 12]
```diff
@@ -0,0 +1,164 @@
...
+                    return redirect('/main/{0}/'.format(object_type))
+                backup.task_id = created_job['id']
+                backup.save()
+                messages.success(request, 'Your backup request has been submitted. \
+                             Once the file is generated we will send you the download\
+                             link on the registered email id.')
+                return redirect('/main/{0}/'.format(object_type))
+            raise ValidationError(('Form data is invalid'), code='invalid')
+
+        except (NotImplementedError, ValidationError) as excp:
```
Comment (aditya999123) [inline apps/backups/views.py:119]: what if there is an exception which is not `NotImplementedError` or `ValidationError` then you would endup returning nothing, this would generate runtime error

File: apps/backups/views.py [hunk 13]
```diff
@@ -0,0 +1,164 @@
...
+            logger.info('Got a notify request from user %s for backup_id: %s',
+                        request.user, backup_id)
+            backup = Backups.objects.get(id=backup_id, user_id__email=request.user)
+            fyle_connection = FyleSdkConnector(backup.fyle_refresh_token)
+            notify_user(fyle_connection, backup.file_path, backup.fyle_org_id,
+                        backup.object_type)
+            messages.success(request, 'We have sent you the download\
+                             link by email.')
+        except Backups.DoesNotExist:
+            messages.error(request, 'You are not authorized to do that!')
```
Comment (aditya999123) [inline apps/backups/views.py:141]: this is not correct, being authorized is different from existence of object

File: apps/backups/views.py [hunk 14]
```diff
@@ -0,0 +1,164 @@
...
+    Home view for Expenses
+    """
+    object_type = 'expenses'
+    def get(self, request):
+        if request.user.refresh_token is None:
+            messages.error(request, 'Please connect your Fyle account!')
+            return redirect('/fyle/connect/')
+        bkp_view = BackupsView()
+        response = bkp_view.get(request, self.object_type)
+        response = json.loads(response.content).get('backups')
```
Comment (aditya999123) [inline apps/backups/views.py:160]: why `.get`

if backups key is not available then response would have value None and backup_list also instead of `[]`

File: apps/data_fetcher/views.py [hunk 15]
```diff
@@ -0,0 +1,74 @@
...
+        fyle_connection = FyleSdkConnector(refresh_token)
+        logger.info('Going to fetch data for backup_id: %s', backup_id)
+        response_data = fyle_connection.extract_expenses(state=filters.get('state'),
+                                                         approved_at=filters.get('approved_at'),
+                                                         updated_at=filters.get('updated_at'))
+        if not response_data:
+            logger.info('No data found for backup_id: %s', backup_id)
+            backup.current_state = 'NO DATA FOUND'
+            backup.save()
+            return HttpResponse(status=200)
```
Comment (aditya999123) [inline apps/data_fetcher/views.py:47]: server should always return a json response

File: apps/data_fetcher/views.py [hunk 16]
```diff
@@ -0,0 +1,74 @@
...
+
+            # Get a secure URL for this backup and mail it to user
+            notify_user(fyle_connection, object_name, fyle_org_id, 'expenes')
+
+            backup.file_path = object_name
+            backup.current_state = 'READY'
+            backup.save()
+            # Remove the files from local machine
+            remove_items_from_tmp(file_path)
+            return HttpResponse(status=200)
```
Comment (aditya999123) [inline apps/data_fetcher/views.py:69]: same as above

File: apps/fyle_connect/utils.py [hunk 17]
```diff
@@ -0,0 +1,44 @@
...
+        self.redirect_uri = settings.REDIRECT_URI
+        self.token_url = settings.TOKEN_URI
+
+    def authorise(self, state):
+        """
+        Initiates the Fyle OAuth2.0 Authorise flow
+        :param state:
+        :return:
+        """
+        authorize_url = self.authorize_uri + '?response_type=code&client_id=' \
```
Comment (aditya999123) [inline apps/fyle_connect/utils.py:27]: please use python library urlib for this
you should avoid string concat

File: apps/fyle_connect/utils.py [hunk 18]
```diff
@@ -0,0 +1,44 @@
...
+        """
+        initialize from settings
+        """
+        self.client_id = settings.CLIENT_ID
+        self.client_secret = settings.CLIENT_SECRET
+        self.authorize_uri = settings.AUTHORIZE_URI
+        self.redirect_uri = settings.REDIRECT_URI
+        self.token_url = settings.TOKEN_URI
+
+    def authorise(self, state):
```
Comment (aditya999123) [inline apps/fyle_connect/utils.py:21]: ple use either authorize
or authorise everywhere

File: apps/backups/views.py [hunk 19]
```diff
@@ -0,0 +1,164 @@
...
+                                                fyle_refresh_token=refresh_token,
+                                                user=user
+                                                )
+
+                # Schedule this backup using JobsInfra
+                fyle_sdk_connector = FyleSdkConnector(refresh_token)
+                fyle_sdk_connection = fyle_sdk_connector.connection
+                jobs = FyleJobsSDK(fyle_sdk_connection)
+                created_job = jobs.trigger_now(
+                    callback_url='{0}{1}'.format(settings.FYLE_JOBS_CALLBACK_URL,
```
Comment (aditya999123) [inline apps/backups/views.py:97]: this can be simplified

File: apps/data_fetcher/utils.py [hunk 20]
```diff
@@ -0,0 +1,263 @@
...
+    """
+    try:
+        message = Mail(
+                        from_email=from_email,
+                        to_emails=to_email,
+                        subject=subject,
+                        html_content=content)
+        sg_client = SendGridAPIClient(settings.SENDGRID_API_KEY)
+        sg_client.send(message)
+    except Exception as e:
```
Comment (aditya999123) [inline apps/data_fetcher/utils.py:240]: why catch this and then raise
and then catch agin on calling function?

File: apps/backups/tests/test_views.py [hunk 21]
```diff
@@ -0,0 +1,42 @@
...
+        self.client.login(email='user1@test.com', password='foo')
+
+    def test_create_backup_failed(self):
+        # name is missing, so form validation should fail
+        post_data = {'approved_at_lte': [''], 'updated_at_gte': [''], 'updated_at_lte': [''],
+                     'download_attachments': ['on'], 'object_type': ['expenses'],
+                     'data_format': ['CSV']}
+        response = self.client.post('/main/backups/', post_data)
+        messages = list(get_messages(response.wsgi_request))
+        self.assertEqual(str(messages[0]), 'Something went wrong. Please try again!')
```
Comment (aditya999123) [inline apps/backups/tests/test_views.py:32]: you should check status_codes not message text

File: apps/backups/tests/test_views.py [hunk 22]
```diff
@@ -0,0 +1,42 @@
...
+
+    def test_create_backup_success(self):
+        post_data = {'name': ['test'], 'approved_at_gte': [''], 'approved_at_lte': [''],
+                     'updated_at_gte': [''], 'updated_at_lte': [''],
+                     'download_attachments': ['on'], 'object_type': ['expenses'],
+                     'data_format': ['CSV']}
+        response = self.client.post('/main/backups/', post_data)
+        messages = list(get_messages(response.wsgi_request))
+        self.assertNotEqual(str(messages[0]), 'Something went wrong. Please try again!')
```
Comment (aditya999123) [inline apps/backups/tests/test_views.py:41]: same

File: apps/backups/views.py [hunk 23]
```diff
@@ -0,0 +1,164 @@
...
+                                                      'current_state',
+                                                      'error_message', 'created_at')[:5]
+        return JsonResponse({"backups": list(backups_list)})
+
+    def post(self, request):
+        try:
+            form = ExpenseForm(request.POST)
+            logger.info('Got a backup request from %s with params: %s',
+                        request.user, request.POST)
+            if form.is_valid():
```
Comment (aditya999123) [inline apps/backups/views.py:73]: write in a a inverse fashion if not valid then raise

you can make your function more readable by saving indents

File: apps/backups/forms.py [hunk 24]
```diff
@@ -0,0 +1,56 @@
...
+    data_format_choices = [
+        ("CSV", "CSV")
+    ]
+    expense_state_choices = [
+        ('FYLED', 'FYLED'),
+        ('PAID', 'PAID'),
+        ('APPROVED', 'APPROVED'),
+        ('DRAFT', 'DRAFT'),
+        ('APPROVER_PENDING', 'APPROVER PENDING'),
+        ('COMPLETE', 'COMPLETE'),
```
[the remark this answers is not in the document]
Comment (arunmcherian94) [reply apps/backups/forms.py:17]: Done.

File: apps/backups/urls.py (hunk 8 above)
Comment (arunmcherian94) [reply apps/backups/urls.py:10]: Done.

File: apps/backups/models.py (hunk 6 above)
Comment (arunmcherian94) [reply apps/backups/models.py:29]: Yup. It get's stored like this :
`{"state": ["FYLED", "PAID", "APPROVED"], "approved_at": [], "updated_at": ["gte:2019-07-01T00:00:00.000Z", "lte:2020-03-28T23:59:59.000Z"], "download_attachments": true}`

File: .env.template (hunk 2 above)
Comment (arunmcherian94) [reply .env.template:8]: Removed. It was added by mistake.

File: apps/backups/models.py (hunk 6 above)
Comment (arunmcherian94) [reply apps/backups/models.py:29]: I will update the help_text for the field. The current one does not explain how it is actually stored.

File: apps/backups/models.py (hunk 3 above)
Comment (arunmcherian94) [reply apps/backups/models.py:23]: Done.

File: apps/backups/models.py (hunk 4 above)
Comment (arunmcherian94) [reply apps/backups/models.py:5]: Done.

File: apps/backups/models.py (hunk 5 above)
Comment (arunmcherian94) [reply apps/backups/models.py:24]: Done.

File: apps/backups/templates/expenses.html (hunk 7 above)
Comment (arunmcherian94) [reply apps/backups/templates/expenses.html:38]: Done.

File: apps/backups/views.py (hunk 12 above)
Comment (arunmcherian94) [reply apps/backups/views.py:119]: Noted. Will add a broader Except block.

File: apps/backups/views.py (hunk 9 above)
Comment (arunmcherian94) [reply apps/backups/views.py:74]: Done.

File: apps/data_fetcher/views.py (hunk 10 above)
Comment (arunmcherian94) [reply apps/data_fetcher/views.py:35]: Done.

File: apps/backups/views.py (hunk 11 above)
Comment (arunmcherian94) [reply apps/backups/views.py:65]: Done.

File: apps/backups/views.py (hunk 13 above)
Comment (arunmcherian94) [reply apps/backups/views.py:141]: Changed the message.

File: apps/data_fetcher/views.py (hunk 15 above)
Comment (arunmcherian94) [reply apps/data_fetcher/views.py:47]: Done.

File: apps/data_fetcher/views.py (hunk 16 above)
Comment (arunmcherian94) [reply apps/data_fetcher/views.py:69]: Done.

File: apps/fyle_connect/utils.py (hunk 18 above)
Comment (arunmcherian94) [reply apps/fyle_connect/utils.py:21]: Done.

File: apps/backups/views.py (hunk 19 above)
Comment (arunmcherian94) [reply apps/backups/views.py:97]: Done.

File: apps/data_fetcher/utils.py (hunk 20 above)
Comment (arunmcherian94) [reply apps/data_fetcher/utils.py:240]: Wanted to bubble this error all the way up to the caller function. The backup succes/failed state gets tracked there.

File: apps/backups/views.py (hunk 23 above)
Comment (arunmcherian94) [reply apps/backups/views.py:73]: Done.

File: fyle_backup_app/settings.py [hunk 25]
```diff
@@ -0,0 +1,242 @@
...
+For more information on this file, see
+https://docs.djangoproject.com/en/3.0/topics/settings/
+
+For the full list of settings and their values, see
+https://docs.djangoproject.com/en/3.0/ref/settings/
+"""
+
+import os
+
+from decouple import config
```
[the remark this answers is not in the document]
Comment (arunmcherian94) [reply fyle_backup_app/settings.py:15]: Done.

File: fyle_backup_app/settings.py [hunk 26]
```diff
@@ -0,0 +1,242 @@
...
+WSGI_APPLICATION = 'fyle_backup_app.wsgi.application'
+
+
+# Database
+# https://docs.djangoproject.com/en/3.0/ref/settings/#databases
+
+DATABASES = {
+    'default': {
+        'ENGINE': 'django.db.backends.mysql',
+        'NAME': config('DB_NAME'),
```
[the remark this answers is not in the document]
Comment (arunmcherian94) [reply fyle_backup_app/settings.py:120]: Done.

File: fyle_backup_app/settings.py [hunk 27]
```diff
@@ -0,0 +1,242 @@
...
+STATIC_ROOT = os.path.join(BASE_DIR, 'staticfiles')
+
+MESSAGE_TAGS = {
+    messages.SUCCESS: 'alert-success',
+    messages.WARNING: 'alert-warning',
+    messages.ERROR: 'alert-danger',
+}
+
+# Fyle OAuth2.0
+BASE_URL = config('BASE_URL')
```
[the remark this answers is not in the document]
Comment (arunmcherian94) [reply fyle_backup_app/settings.py:178]: Done.

File: .env.template (hunk 1 above)
Comment (arunmcherian94) [reply .env.template:3]: Done.
